### ISSUE TITLE
Compiling SOWFA - horizontalAxisWindTurbinesALM.H & hdf5.h

### ISSUE THREAD

**Issue by ColinB07** [pablo-benito/SOWFA-installation#2]
Hi Pablo,

I tried to compile SOWFA with OpenFAST, but I'm encountering multiple fatal errors for missing files.

First, the OpenFAST interface compilation breaks.
`/home/breiffni/OpenFAST-build//include/SC.h:5:10: fatal error: hdf5.h: No such file or directory
    5 | #include "hdf5.h"
      |          ^~~~~~~~
compilation terminated.`

then `windPlantSolver.C:59:10: fatal error: horizontalAxisWindTurbinesALM.H: No such file or directory
   59 | #include "horizontalAxisWindTurbinesALM.H"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.`
{{EDIT: Solved with `export SOWFA_DIR="/path/to/SOWFA`}} 

I'm on Ubuntu 18.04, gcc 10.0.1.
Here is the full log : [logmakeSOWFA.txt](https://github.com/pablo-benito/SOWFA-installation/files/4237118/logmakeSOWFA.txt)

Any clue would be very grateful. Thank you for your time and your patience.

**Comment by pablo-benito** [comment]
Basically, your compiler cannot find the HDF5 header files (hdf5.h), needed to compile this part of SOWFA. 

As you can see in the document, first you have to install the development package of the HDF5 library ( libhdf5-dev ), and then you have to edit the `options` file, inside the folder:

`applications/solvers/incompressible/windEnergy/pisoFoamTurbine.ALMAdvancedOpenFAST/Make/
`

to add this `-I$(HDF5_DIR)/include`, to the `EXE_INC` variable.

Then, before launching the compilation, you have to export the HDF5_DIR variable pointing to the location of the HDF5 library. On Ubuntu 18.04 this path must be

 `/usr/lib/x86_64-linux-gnu/hdf5/serial`

Hope it helps

**Comment by ColinB07** [comment]
I've done these steps, with no further improvements.
I tried to locate the file with `sudo find / -iname "hdf5.h"`. It has been found here `/usr/include/hdf5/serial/hdf5.h`. 
So I've tried again `./Allwmake` after `export HDF5_DIR="/usr/include/hdf5/serial/"`, same issue.
I've also tried to use gcc/g++ 5, but it does not change anything.

**Comment by pablo-benito** [comment]
yes, the 'real' location of the file is inside` /usr/include/hdf5/serial/hdf5.h`, but is easier to use the convenience folder `/usr/lib/x86_64-linux-gnu/hdf5/serial`, which has an 'include' and a 'lib' folder inside. 

By using this path as `HDF5_DIR`, you can define with  only one variable, the location of the headers and the libraries of HDF5 in the options file: 
The location of the headers, will be` -I$(HDF5_DIR)/include` (you have to add this line as mentioned in my doc!! )
 and  the location of the HDF5 libraries is already defined in the EXE_LIBS as `-L$(HDF5_DIR)/lib`

If you point the `HDF5_DIR` variable to the path you mentioned, the include path passed to the compiler would be /usr/include/hdf5/serial/include, that doesn't even exist.. 


TL;DR;
- Inside the `options` file, of `applications/solvers/incompressible/windEnergy/pisoFoamTurbine.ALMAdvancedOpenFAST/Make/`, add -I$(HDF5_DIR)/include at the end of the EXE_INC variable,  
- export HDF5_DIR="/usr/lib/x86_64-linux-gnu/hdf5/serial"
- Recompile

**Comment by ColinB07** [comment]
I'm feeling really dumb, because these steps are not working for me.

Step 1: [options.txt](https://github.com/pablo-benito/SOWFA-installation/files/4244330/options.txt)
Step 2: 
`~/SOWFA$ export HDF5_DIR="/usr/lib/x86_64-linux-gnu/hdf5/serial"
~/SOWFA$ WM_PROJECT_USER_DIR=${PWD} ./Allwmake`

Results : 
`In file included from /home/breiffni/OpenFAST-build/include/OpenFAST.H:20,
                 from horizontalAxisWindTurbinesALMOpenFAST/horizontalAxisWindTurbinesALMOpenFAST.H:68,
                 from horizontalAxisWindTurbinesALMOpenFAST/horizontalAxisWindTurbinesALMOpenFAST.C:36:
/home/breiffni/OpenFAST-build/include/SC.h:5:10: fatal error: hdf5.h: No such file or directory
    5 | #include "hdf5.h"
      |          ^~~~~~~~
compilation terminated.
`

Full log [logmakeSOWFA1.txt](https://github.com/pablo-benito/SOWFA-installation/files/4244357/logmakeSOWFA1.txt)

**Comment by pablo-benito** [comment]
How, it seems that SOWFA source code changed a little bit since last time I checkit out... 

Rigth now, the inclusion of the HDF5 comes from an extra place... 
Go to `src/turbineModels/turbineModelsOpenFAST/Make/options` and add the following lines:

`-I$(HDF5_DIR)/include` at the end of the EXE_INC variable

and  

```
-L$(HDF5_DIR)/lib/ \
    -lhdf5_hl \
    -lhdf5
```

at the end of the EXE_LIBS and try again...

**Comment by ColinB07** [comment]
Thank you very much for your patience and your patience, this problem is now solved !

However, a new error occurs :
`/usr/bin/ld: warning: libversioninfolib.so, needed by /home/breiffni/OpenFAST-build/lib/libopenfastcpplib.so, not found (try using -rpath or -rpath-link)
/usr/bin/ld: warning: libgfortran.so.4, needed by //usr/lib/x86_64-linux-gnu/libopenblas.so.0, may conflict with libgfortran.so.5
/home/breiffni/OpenFAST-build/lib/libopenfast_postlib.so: undefined reference to '__versioninfo_MOD_querygitversion'
collect2: error: ld returned 1 exit status`
May I ask you again some help about it?


[About the missing 'horizontalAxisWindTurbinesALM.H', in the same option file, the environment  SOWFA_DIR was not defined. So the compiler was not able to reach the file. I solved it with `export SOWFA_DIR="/path/to/SOWFA"`]

**Comment by pablo-benito** [comment]
hummm.. 

> /usr/bin/ld: warning: libgfortran.so.4, needed by //usr/lib/x86_64-linux-gnu/libopenblas.so.0, may conflict with libgfortran.so.5 /home/breiffni/OpenFAST-build/lib/libopenfast_postlib.so

Did you compile OpenFAST with a different compiler version from the default one that comes with Ubuntu?? Because it seems that the OpenFAST libraries are compiled with a newer gfortran version (libgfortran.so.5), and looking at your first post, it looks that you're using gcc/gfortran 10

If that's the case you can have problems, as some of the Fortran libraries that you're using (like OpenBLAS, etc) are already compiled with the default ubuntu gfortran version, that is version 7.4.0, and uses an older version of the gfortran library (libgfortran.so.4). So you're mixing Fortran code compiled with different compilers., and that is definitely a bad idea (Fortran do not have an ABI standard.. bla bla bla). 

So, unless you have a very good reason to use a newer version of the gcc/gfortran compiler, I recomend you to use the default one. That is, recompile OpenFAST/OpenFOAM/SOWFA with the default gcc/gfortran version... Sorry, but the other option is to rebuild everything with the newer version, including the openblas library, which is a pain in the ass...

**Comment by ColinB07** [comment]
I tried to start again the compilation, from new archive file after resetting the terminal. However, by default, the compilation is using libgfortran.so.5. I know I can change gcc compiler with `export WM_CC='gcc-version'`, how can I change the fortran compiler?

**Comment by pablo-benito** [comment]
OpenFoam (and SOWFA) is C/C++ only, and there is no variable to set a Fortran compiler. But, you will use gcc not only as compiler but as a linker, so again, you can avoid a lot of weird errors if you use the same gcc version as the gfortran compiler.  

It's OpenFAST the one that is Fortran based. And is OpenFAST the one that is linking to the system openBLAS library, which, in turn is linked to the system libgfortran.so.4. But is the rest of the OpenFAST code the one that is linking to the libgfortran.so.5, because you're using a different compiler version. That's the main cause of your errors, you have a piece of code linked to different versions of the same library.

So first of all, you must reconfigure from scratch your OpenFAST compilation, telling cmake to use the system versions of gcc/gfortran.  During the cmake configuration phase, you will see the version of the compilers that is going to use. Then compile and install OpenFAST as usual.
 
After that, you must recompile OpenFOAM, again, using the system gcc. And, then, finally, recompile SOWFA

Carry on! you're really close ;)

**Comment by pablo-benito** [comment]
@ColinB07 , 
yesterday I compiled myself the whole SOWFA code from scratch and I detected another issue:
OpenFAST added a couple of months ago another library, libversioninfolib, that must be included in the dependencies... 

I've updated my document with all my recent findings, so please check it out.

**Comment by ColinB07** [comment]
Thank you very much for your help !

**Comment by ColinB07** [comment]
I've compiled again everything, no issue, no flag, perfect !
So, I've tried the first example : example.ABL.flatTerrain.neutral.
No alarming flags for the preprocess, however the solving part is not running at all.
The error is : 
`--------------------------------------------------------------------------
mpirun was unable to find the specified executable file, and therefore
did not launch the job.  This error was first reported for process
rank 0; it may have occurred for other processes as well.

NOTE: A common cause for this error is misspelling a mpirun command
      line parameter option (remember that mpirun interprets the first
      unrecognized command line token as the executable).

Node:       breiffni-XPS-8900
Executable: setFieldsABL
--------------------------------------------------------------------------
8 total processes failed to start`
I didn't find any similar issue for SOWFA anywhere. May you have an idea of the root of the problem?

~I don't know if the following information are relevant:
>>which mpirun
/home/breiffni/OpenFOAM/OpenFOAM-2.4.x/bin/mpirun

(in bashrc file) export WM_MPLIB=SYSTEMOPENMPI



logs:

[log1ABLSolver.txt](https://github.com/pablo-benito/SOWFA-installation/files/4255679/log1ABLSolver.txt)
[log1setFieldsABL.txt](https://github.com/pablo-benito/SOWFA-installation/files/4255680/log1setFieldsABL.txt)
[logblockMesh.txt](https://github.com/pablo-benito/SOWFA-installation/files/4255681/logblockMesh.txt)
[logcheckMesh1.txt](https://github.com/pablo-benito/SOWFA-installation/files/4255682/logcheckMesh1.txt)
[logdecomposePar.txt](https://github.com/pablo-benito/SOWFA-installation/files/4255683/logdecomposePar.txt)
[logrenumberMesh.txt](https://github.com/pablo-benito/SOWFA-installation/files/4255684/logrenumberMesh.txt)

**Comment by pablo-benito** [comment]
> I've compiled again everything, no issue, no flag, perfect !

Great!

> No alarming flags for the preprocess, however the solving part is not running at all.

The error that you see it's basically that mpirun cannot find the SOWFA binary that you want to launch.
I mean, is basically that the ABLSolver, and the setFieldsABL, etc,  are not in a directory included in the PATH environment variable. So your shell is not able to locate the program, and mpirun neither.

You can add the SOWFA bin directory to the path, and then launch the command:
```bash 
export PATH=${SOWFA_DIR}/applications/bin/${WM_OPTIONS}:${PATH}
ABLSolver  

# or, if you want to launch it in parallel, in 4 cores, for example
mpirun -np 4 ABLSolver -parallel  
``` 
another option is  launching the command using the full path to the program:
```bash 
${SOWFA_DIR}/applications/bin/${WM_OPTIONS}/ABLSolver

# or, if you want to launch it in parallel, in 4 cores, for example
mpirun -np 4 ${SOWFA_DIR}/applications/bin/${WM_OPTIONS}/ABLSolver -parallel 
```

But, before launching in parallel, try to launch it first in a single core, just to check that everything is ok. Also make sure that the binary is able to locate _ALL_ the libraries that depends on, otherwise you will see errors like "Error while loading shared libraries...".

Check the [ _Runtime configuration_ ](https://github.com/pablo-benito/SOWFA-installation#runtime-configuration ) of my doc for this.

**Comment by ColinB07** [comment]
Thank you very much for your patience !

I still have an issue with ABLSolver: `mpirun noticed that process rank 6 with PID 0 on node breiffni-XPS-8900 exited on signal 9 (Killed).` But it seems to be a memory error, I should be able to deal with it.

Thanks again for your time!

**Comment by pablo-benito** [comment]
Signal 9, is sigkill, it can be virtually anything.. 
No other log? anything? 

Did the other commands of the preprocess run succesfully?

I had similar errors some time ago, when I compiled OpenFOAM 2.4.x with a gcc  version 6 or higher: The original code of OpenFOAM 2.4.x was not 100% ok, it did compile, but starting with that version of gcc, the code compiles, but the resulting binaries produced this kind of weird errors. That's why I have my own repository for OpenFOAM 2.4.x, I had to patch the OpenFoam code to properly support modern compilers...

In Ubuntu 18.04 the gcc version is 7.4.0, so if you downloaded the OpenFOAM code from the official github site, you will hit this error.. And then you should download my OpenFOAM code, and recompile.. :(

**Comment by ColinB07** [comment]
I just have these warnings : 
`--> FOAM Warning : 
    From function dlOpen(const fileName&, const bool)
    in file POSIX.C at line 1179
    dlopen error : /home/breiffni/SOWFA//lib/linux64GccDPOpt/libSOWFAincompressibleLESModels.so: undefined symbol: _ZTIN4Foam14incompressible9LESModels11GenEddyViscE
--> FOAM Warning : 
    From function dlLibraryTable::open(const fileName&, const bool)
    in file db/dynamicLibrary/dlLibraryTable/dlLibraryTable.C at line 96
    could not load "libSOWFAincompressibleLESModels.so"`
which seem to be safe (https://github.com/NREL/SOWFA/issues/25)

The program is running perfectly until the sudden error, the memory outflow issue seemed plausible.
[logABLSolver.txt](https://github.com/pablo-benito/SOWFA-installation/files/4262980/logABLSolver.txt)

I did the set-up from your code, but I'll try recompile one more time.

**Comment by pablo-benito** [comment]
Hum, I did a test with the binaries that I compiled myself yesterday, and the example  example.ABL.flatTerrain.neutral ran perfectly... 

I only had  to change the number of processors in the blockMesh dict and in the setUp file to 4 so that the example can run in my laptop, and everything went ok.. 

I dont know why the example defines the number of processors in different places..

**Comment by ColinB07** [comment]
The same issue comes back after recompiling. Adjusting the processors do not solve the issue. I did not adjust the number of cells, I'll try this. 
A friend of mine is going to check my set-up next week otherwise.

Thank you again so much for your patience.

**Comment by ColinB07** [comment]
The problem was the number of cells. It works now. Thank you !

**Comment by pablo-benito** [comment]
You're welcome, it's always a pleasure. 
Have fun :) 

I close this issue.

**Comment by csq1234** [comment]
Hello, pablo.
I tried to run ALM , I'm going on ALM example debugging, I've flatTerrain. Neutral data import ALM example, and draw the grid successfully, but met meet a problem in the next step . 
The log.1. WindPlantSolver. ALM file:

- -&gt; FOAM FATAL TO ERROR

- :cannot find file

- File: / home/huyang/OpenFOAM/huyang - 2.4 x/my_cases ALM9 / processor13/20000 / T at line 0.From function regIOobject::readStream()in file db/regIOobject/regIOobjectEead.C at line 73FOAM parallel run exiting

- If you have encountered similar problems. Your reply will be highly appreciated.

**Comment by pablo-benito** [comment]
Well, first of all, I'm not a SOWFA expert, I was just in charge of compiling it! 

Second point. This issue is closed, and your issue is not even related with the title of the original issue! Please open a new one next time..  This will help me, and others trying to find an answer to a similar problem.

Third point, your issue is not related with compiling & installing SOWFA, but rather is an issue running the software. The first point to ask a question like this is the official repo of SOWFA. 

Having said that, you must explain a little more your issue: What have you done exactly, what commands have you executed, what files have you modified, etc..  
The error is quite clear: OpenFOAM is not finding that file `/ home/huyang/OpenFOAM/huyang-2.4x/my_casesALM9/processor13/20000/T` 
It seems that you have decomposed the domain in less than 13 cores, and then you are trying to execute it in 13 or more cores.

A part from that. One of the main problems with the SOWFA examples is that they are intended to be executed inside the NREL supercomputing facilities. The scripts that they put in the examples only works there, in a supercomputing facility with access to a PBS queue system, 'modules' and so on. So, outside NREL, you have to read those scripts, understand them, and create your own scripts to execute something similar on your computer or your facility. 

But, if you look closely to some of the 'examples', you will find that some of them requires a LOT of cores to run. For example, the script for running example.NREL5MW.ALMAdvancedFAST8 try to decompose & run in 960 cores! If you are just testing SOWFA in your laptop, be sure to adjust the example to something that fits your hardware..

**Comment by csq-123** [comment]
I'm sorry for my mistake and I'll pay attention next time.
My server has 16 cores and I changed the cores number to 16 in the runscript.preprocess,runscript.solve.1 and setUp. I also changed decompOrder to (4 2 2) in the setUp file. I mainly changed these variables and the other parameters were default. Now, I've imported the flatTerrain. Neutral data to the ALM example by modifying the precursorDir path, and draw the grid successfully, but met a problem in the next step to run runscript.solve.1.
